### ISSUE TITLE
metamorphic: TestMeta failed

### ISSUE THREAD

**Issue by nicktrav** [cockroachdb/pebble#2036]
This failure wasn't reported due to a hung build.

```console
[07:04:04] :     [Step 1/1] ===== SEED =====
[07:04:04] :     [Step 1/1] 1665903830119504942
[07:04:04] :     [Step 1/1] ===== DIFF =====
[07:04:04] :     [Step 1/1] /artifacts/meta/221016-070350.119/{standard-000,random-021}
[07:04:04] :     [Step 1/1] @@ -882,14 +882,14 @@
[07:04:04] :     [Step 1/1]  batch8.Set("qjxnplqtjw@8", "egmyyazv") // <nil> #881
[07:04:04] :     [Step 1/1]  iter8.Next("") // [false] <nil> #882
[07:04:04] :     [Step 1/1]  db.Merge("nyexd@1", "mysbbqesmnpsw") // <nil> #883
[07:04:04] :     [Step 1/1]  iter8.Next("") // [false] <nil> #884
[07:04:04] :     [Step 1/1]  iter8.SetBounds("ptsrrfqypv@3", "yjijntpmaka@6") // <nil> #885
[07:04:04] :     [Step 1/1] -iter8.SeekGE("ptsrrfqypv@3", "") // [true,"qnwxpsrcb@7","iupkbmqd",<no range>] <nil> #886
[07:04:04] :     [Step 1/1] -iter8.Next("") // [true,"tlnqgaky@7","",<no range>] <nil> #887
[07:04:04] :     [Step 1/1] +iter8.SeekGE("ptsrrfqypv@3", "") // [true,"tlnqgaky@7","",<no range>] <nil> #886
[07:04:04] :     [Step 1/1] +iter8.Next("") // [true,"vzpbhpufrfp",<no point>,["vzpbhpufrfp","wwazc")=>{"@7"="ikoye"}*] <nil> #887
[07:04:04] :     [Step 1/1]  batch8.Merge("jpqbmi@5", "qprvtoqx") // <nil> #888
[07:04:04] :     [Step 1/1] -iter8.Next("") // [true,"vzpbhpufrfp",<no point>,["vzpbhpufrfp","wwazc")=>{"@7"="ikoye"}*] <nil> #889
[07:04:04] :     [Step 1/1] +iter8.Next("") // [false] <nil> #889
[07:04:04] :     [Step 1/1]  iter8.SeekPrefixGE("purvnwuo@3") // [false] <nil> #890
[07:04:04] :     [Step 1/1]  db.DeleteRange("gknmriqj@7", "jvoid") // <nil> #891
[07:04:04] :     [Step 1/1]  iter8.SeekPrefixGE("tgbxszeqkgx@3") // [false] <nil> #892
[07:04:04] :     [Step 1/1]  db.Get("wwazc@5") // ["puglsqifetkfnxbkvld"] <nil> #893
[07:04:04] :     [Step 1/1]  db.Merge("ptsrrfqypv", "uldgqcyvfrnglttgv") // <nil> #894```

**Comment by nicktrav** [comment]
This reproduces locally:

```bash
$ go test -tags 'invariants' -exec 'stress -p 1' -timeout 0 -test.v -run TestMeta$ ./internal/metamorphic -seed 1665903830119504942 -ops "uniform:5000-10000"
```

**Comment by jbowens** [comment]
The random-021 run iter8 is not observing the key-value pair
```
"qnwxpsrcb@7#234,1", "iupkbmqd"
```
That key-value pair is within the iterator (`iter8`)'s time-bound iterator bounds (`[6,9)`) and was committed (at operation 762) before iter8's backing snapshot (`snap6`) was created (at operation 820).

The merging iterator never observes the key:
```
 seeked level 0 (*pebble.levelIter) to "ptsrrfqypv@3": (<nil>, "")
 seeked level 1 (*pebble.levelIter) to "ptsrrfqypv@3": ("yjijntpmaka@6#72057594037927935,15", "")
 seeked level 2 (*pebble.levelIter) to "ptsrrfqypv@3": ("yjijntpmaka@6#72057594037927935,15", "")
    2 tombstone owdmqfn@5-puhvnwihpk:{(#216,RANGEDEL) (#158,RANGEDEL)} contains seek key "ptsrrfqypv@3"; adjusting seek key to "puhvnwihpk"
 seeked level 3 (*pebble.levelIter) to "puhvnwihpk": ("yjijntpmaka@6#72057594037927935,15", "")
    3 tombstone ptsrrfqypv-vybvotm@2:{(#144,RANGEDEL)} contains seek key "puhvnwihpk"; adjusting seek key to "vybvotm@2"
 seeked level 4 (*pebble.levelIter) to "vybvotm@2": (<nil>, "")
mergingIter.SeekGE("ptsrrfqypv@3", TrySeekUsingNext()=false) = (<nil>, "")
```

The top level was seeked to file:
```
 seeked level 1 (*pebble.levelIter) to "ptsrrfqypv@3": ("yjijntpmaka@6#72057594037927935,15", "")
    levelIter.SeekGE("ptsrrfqypv@3") seeked to file 000027:[aywfrtm@1#203,DEL-zfbdegncf#179,MERGE] points:[aywfrtm@1#203,DEL-zfbdegncf#179,MERGE] ranges:[baomyybcov#201,RANGEKEYSET-wwazc#72057594037927935,RANGEKEYSET] (seqnums: [152-216])
```
This file's largest sequence number (216) is less than the missing key's sequence number (234). Are we missing a recently flushed file?

The most recent flush was also the first flush after the missing key was committed. It produced sstable 000030:
```
// INFO: [JOB 6] ingesting: sstable created 000028
// INFO: [JOB 7] WAL created 000029
// INFO: [JOB 8] flushing 1 memtable to L0
// INFO: [JOB 8] flushing: sstable created 000030
// INFO: [JOB 8] flushed 1 memtable to L0 [000030] (2.1 K), in 0.0s (0.0s total), output rate 8.1 M/s
// INFO: [JOB 8] WAL deleted 000026
// INFO: ingest failed to remove original file: invalid argument
// INFO: [JOB 6] ingested L0:000028 (998 B)
```
why didn't we read it?

[posting notes and will pick back up here]

**Comment by jbowens** [comment]
It might be another interaction between the monotonic bounds optimization and block-property filters?

The seek within the relevant file hit the single-level iterator fast path for monotonically increasing bounds, but the index was already positioned beyond the key we were supposed to return. 
```
    levelIter.SeekGE("ptsrrfqypv@3") seeked to file 000030:[bimqcfpaq#244,RANGEKEYDEL-zfbdegncf#218,DEL] points:[gknmriqj@7#240,SET-zfbdegncf#218,DEL] ranges:[bimqcfpaq#244,RANGEKEYDEL-vzpbhpufrfp#72057594037927935,RANGEKEYDEL] (seqnums: [218-248])
          twoLevelIterator.SeekGE("ptsrrfqypv@3")
          dontSeekWithinSingleLevelIter false
          singleLevelIterator.SeekGE("ptsrrfqypv@3")
          singleLevelIterator.SeekGE("ptsrrfqypv@3"); attempting fast path (boundsCmp = 1); index block currently positioned at "zfbdegncf#218,0"
```

**Comment by jbowens** [comment]
Instrumenting the metamorphic tests to stop after the relevant operation, 000030.sst does indeed contain the relevant key:

```
000030.sst
gknmriqj@7#240,SET [756d68716566756763656a6f7665]
ikhvwbwg@3-jvoid#237,RANGEDEL
ioxx@3#247,MERGE [716a77]
jqfifngjn@5#245,SET [7366796177726d7773]
jvoid-lvncvmt#237,RANGEDEL
lvncvmt-lvncvmt@1#232,RANGEDEL
lvncvmt@1-mkrqulsg@6#232,RANGEDEL
mkrqulsg@6-ompi@4#232,RANGEDEL
nwkp@5#248,MERGE [6c766a6f]
ompi@4-ptsrrfqypv@3#224,RANGEDEL
ompi@2#231,DEL []
ptsrrfqypv@3#230,MERGE [62647563666a667671796c787961776a7465]
purvnwuo@3#239,MERGE [706c646a6f647a6f6676617972]
qnwxpsrcb@7#234,SET [6975706b626d7164]
tgbxszeqkgx@9#246,SET [7877676d75716765]
vybvotm#243,DEL []
wwazc@5#242,MERGE [64]
wzaj@3#236,DEL []
yjijntpmaka@5#227,SET [73697764756c70]
zfbdegncf#218,DEL []
[bimqcfpaq-idcungxoc):
  #244,RANGEKEYDEL
[idcungxoc-nyexd):
  #244,RANGEKEYDEL
[nyexd-purvnwuo):
  #219,RANGEKEYUNSET: @3
[purvnwuo-vzpbhpufrfp):
  #241,RANGEKEYDEL
```

**Comment by jbowens** [comment]
Disabling the two-level iterator's SeekGE boundsCmp optimization gets us to the correct key.

**Comment by jbowens** [comment]
The top-level index entry `"yjijntpmaka@5#227,1"` is being skipped and instead we're jumping to top-level index entry `zfbdegncf#218,0`.

Looking back at the last time this sstable iterator was positioned was during `SeekGE("nyexd@2")`:

The iterator initially seeked within the two-level iterator, finding the top-level index entry that we will later skip when looking for `qnwxpsrcb@7`: `i.topLevelIndex.Key() = "yjijntpmaka@5#227,1"; seek key "nyexd@2"; boundsCmp = 0`. When the iterator falls down to the single-level iterator, it finds the iterator already exhausted, performing a noop: `          singleLevelIterator.SeekGE("nyexd@2"); already exhausted so no-op (i.index.Key() = "pu\\x00#72057594037927935,17", i.exhaustedBounds=0, i.data.isDataInvalidated()=true)`. This exhaustion claims to be not have been because the bounds were exceeded, but because we truly iterated to the end of the block:
```
	// The i.exhaustedBounds comparison indicates that the upper bound was
	// reached. The i.data.isDataInvalidated() indicates that the sstable was
	// exhausted.
```

But the seek on this sstable iterator before that exhausted bounds:
```
seeked level 0 (*pebble.levelIter) to "jqfifngjn@5": ("kjpe@1#72057594037927935,15", "")
    levelIter.SeekGE("jqfifngjn@5") seeked to file 000030:[bimqcfpaq#244,RANGEKEYDEL-zfbdegncf#218,DEL] points:[gknmriqj@7#240,SET-zfbdegncf#218,DEL] ranges:[bimqcfpaq#244,RANGEKEYDEL-vzpbhpufrfp#72057594037927935,RANGEKEYDEL] (seqnums: [218-248])
          twoLevelIterator.SeekGE("jqfifngjn@5")
        BlockPropertyFilter[filterInterval=[6,9)].Intersects([0,18446744073709551615)) = true
          twoLevelIterator: i.topLevelIndex.Key() = "yjijntpmaka@5#227,1"; seek key "jqfifngjn@5"; boundsCmp = 0
          dontSeekWithinSingleLevelIter false
          singleLevelIterator.SeekGE("jqfifngjn@5", TrySeekUsingNext()=false)
          singleLevelIterator.SeekGE("jqfifngjn@5"); slow path
          singleLevelIterator.SeekGE("jqfifngjn@5"); i.index.SeekGE("jqfifngjn@5") = "k\\x00#72057594037927935,17" path
        BlockPropertyFilter[filterInterval=[6,9)].Intersects([3,6)) = false
        BlockPropertyFilter[filterInterval=[6,9)].Intersects([2,6)) = false
        BlockPropertyFilter[filterInterval=[6,9)].Intersects([3,4)) = false
          twoLevelIterator.skipForward:
          i.exhaustedBounds > 0: returning nil, nil (i.topLevelIndex.Key() = "yjijntpmaka@5#227,1", i.index.Key() = "pu\\x00#72057594037927935,17")
```

The previous bounds of the iterator were:
```
iter8.SetBounds("jqfifngjn@5", "ptsrrfqypv@3") // <nil> #852
```
and then they moved to
```
iter8.SetBounds("ptsrrfqypv@3", "yjijntpmaka@6") // <nil> #885
```

**Comment by jbowens** [comment]
tldr: The crux of the issue seems to be that an iterator with a block-property filter may leave the single-level iterator with `i.data.isDataInvalidated()=true` when it exhausts the bounds. If the bounds are later modified, a subsequent seek may improperly interpret the `i.data.isDataInvalidated()=true` and `i.boundsExhausted=0` as the entire index block being exhausted.

@sumeerbhola — do you have thoughts on the right way to fix this?

----

To summarize we have 5 iterator operations of interest:

1. `iter8.SetBounds("jqfifngjn@5", "ptsrrfqypv@3") // <nil> #852`
2. `iter8.SeekGE("alwyfah", "") // [true,"nyexd",<no point>,["nyexd","ptsrrfqypv@3")=>{"@7"="o"}*] <nil> #879`
3. `iter8.SeekGE("nyexd@2", "") // [false] <nil> #880`
4. `iter8.SetBounds("ptsrrfqypv@3", "yjijntpmaka@6") // <nil> #885`
5. `iter8.SeekGE("ptsrrfqypv@3", "") // [true,"qnwxpsrcb@7","iupkbmqd",<no range>] <nil> #886`

----

Working backwards:

5\. This SeekGE improperly skips the index block with key `"yjijntpmaka@5#227,1"`, because the bounds were monotonically moved forward and the two-level iterator is already positioned at `"zfbdegncf#218,0"`. Why is the two-level iterator already positioned at `"zfbdegncf#218,0"`?

4\. This SetBounds set `boundsCmp=+1` since the bounds did move forward monotonically.

3\. This previous seek hit the two-level index fast path, finding it was already positioned at `"yjijntpmaka@5#227,1"`. The call into the singleLevelIterator hit the no-op fast path. The index block was positioned at `i.index.Key() = "pu\\x00#72057594037927935,17"`. The bounds were not marked as exhausted `i.exhaustedBounds=0`, but the data block had been invalidated `i.data.isDataInvalidated()=true`, which the no-op fast path interprets as exhaustion of the index block. This caused the two-level iterator to skip forward to `"zfbdegncf#218,0"`. Why was the single-level iterator's data block invalidated when there were additional data blocks (including the one keyed by ` in the first-level index?

2\. This previous seek performed a slow-path two-level index seek, finding the index block keyed by `"yjijntpmaka@5#227,1"`. It hit the slow-path in the first-level index seek too. It then skipped blocks `"k\\x00#72057594037927935,17"`, `"p\\x00#72057594037927935,17"` and `"pu\\x00#72057594037927935,17"` as irrelevant due to the block-property filter. Then it marked bounds as exhausted since `pu\\x00` is greater than the upper bound `ptsrrfqypv@3`. No data block was ever loaded, so `i.data.isDataInvalidated()=true`.

**Comment by sumeerbhola** [comment]
(flushing the comment that I was in the middle of writing before our conversation)
I don't understand step 3. I assume this had `flags.TrySeekUsingNext()`. So it should fall into this fast-path and call `singleLevelIterator.SeekGE`
https://github.com/cockroachdb/pebble/blob/f34af25a01873c9cf027e53147754e96ec8b2b53/sstable/reader.go#L1554-L1574
which should return here because of `exhaustedBounds==+1`
https://github.com/cockroachdb/pebble/blob/f34af25a01873c9cf027e53147754e96ec8b2b53/sstable/reader.go#L609-L612
so when the twoLevelIterator falls through to https://github.com/cockroachdb/pebble/blob/f34af25a01873c9cf027e53147754e96ec8b2b53/sstable/reader.go#L1889-L1891
it should return
------
As we realized, the problem is that we have made `singleLevelIterator` forget that it has exhaustedBounds=+1, which doesn't make it do work in this case since `i.data.isDataInvalidated()` is also true (but in some other cases could). And then we come back and incorrectly skip forward in `twoLevelIterator.skipForward` to the next top-level index entry.
